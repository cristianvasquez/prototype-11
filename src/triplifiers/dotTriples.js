// This should be done at some point with
// https://chevrotain.io/docs/ (if worth the effort)

import { extract } from './miniNLP.js'
import { applyToChunks } from './utils.js'

//  (un :: (texto :: bonito))
const regexpSPO = /\(\s*([^)]+)\s*::\s*\(([^)]+)\s*::\s*([^)]+)\s*\)\)/g
//  un :: (texto :: bonito)
const regexpSPOnp = /([^\n\\)]+)\s*::\s*\(([^)]+)\s*::\s*([^)]+)\s*\)/g
// (texto :: bonito)
const regexpPO = /\(([^)]+)\s*::\s*([^)]+)\s*\)/g
// texto :: bonito
const regexpPOnp = /([^\n]+)::([^\n]+)/g

function trim (txt) {
  return txt.replace(/^\s+|\s+$/gm, '')
}

function toSPO (text) {
  const result = []

  // Matches SPO with parenthesis
  for (const match of text.matchAll(regexpSPO)) {
    result.push({
      subject: toTerm(match[1]),
      predicate: toTerm(match[2]),
      object: toTerm(match[3])
    })
  }
  text = text.replace(regexpSPO, '')

  // Matches SPO without parenthesis
  for (const match of text.matchAll(regexpSPOnp)) {
    result.push({
      subject: toTerm(match[1]),
      predicate: toTerm(match[2]),
      object: toTerm(match[3])
    })
  }
  text = text.replace(regexpSPOnp, '')

  // Matches PO with parenthesis
  for (const match of text.matchAll(regexpPO)) {
    result.push({
      predicate: toTerm(match[1]),
      object: toTerm(match[2])
    })
  }
  text = text.replace(regexpPO, '')

  // Matches PO without parenthesis
  for (const match of text.matchAll(regexpPOnp)) {
    result.push({
      predicate: toTerm(match[1]),
      object: toTerm(match[2])
    })
  }
  return result
}

function toTerm (value) {
  const trimmed = trim(value)
  const tryArray = trimmed.split(',')
  if (tryArray.length > 1
    && trimmed.startsWith('[')
    && trimmed.endsWith(']')) {
    return {
      value: trimmed.slice(1, -1).split(',').map(toTerm),
    }
  }

  return {
    value: trimmed,
    entities: extract(value)
  }
}


function * getDotTriples (text) {
  for (const triples of applyToChunks(text, toSPO)) {
    if (triples) {
      yield triples
    }
  }
}

export { getDotTriples, toSPO }