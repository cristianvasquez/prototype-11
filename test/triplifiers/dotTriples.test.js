import { getDotTriples } from '../../src/triplifiers/dotTriples.js'

import expect from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'

expect.extend({ toMatchSnapshot })


const phrases = [
  'properties :: and values are separated by double colons',
  `
  to simplify things, this
  mini-syntax :: spans
  only one line
  `,
  'Maria :: climbed :: the tree',
  'subject :: \'Quotes escape :: characters\' :: object',
  'internal link :: [[link]]',
  'is a :: [[Editor]], [[Jour]]',
  'external link :: <http://example.org>',
  'external link :: http://example.org',
  'date :: 2022-01-07 15:07',
  'arrays :: [banana, apple]',
  'A :: B :: C :: D',
  'vaccination dates :: 2022-01-07, next week',
  '__THIS__ :: is a :: note',
  ' something :: points at :: __THIS__',

  // '[Maria, Bob] :: (climbed :: the tree)',
  // '[Maria, Bob] :: ([climbed,looked at] :: the tree)',
  // '[Maria, Bob] :: ([climbed,looked at] :: [the tree, the sun])',
  // 'That night,  (Bob :: (ate :: all the pancakes)), and (we :: (are still :: surprised))',
  // 'Bob :: (has :: drivers license), (can drive :: true)',
]



describe('[getDotTriples]', function () {

  phrases.forEach((current)=>{
    it(`"${current}"`, async function () {

      let actual = []
      for (const triples of getDotTriples(current)) {
        actual = [...actual,...triples]
      }
      // console.debug(JSON.stringify(actual,null,2))
      expect(actual).toMatchSnapshot(this)
    })
  })

})