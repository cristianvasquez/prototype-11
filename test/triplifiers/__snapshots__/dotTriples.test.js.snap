// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[getDotTriples] "
  to simplify things, this
  mini-syntax :: spans
  only one line
  " 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "spans",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "mini-syntax",
    },
  },
]
`;

exports[`[getDotTriples] " something :: points at :: __THIS__" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "__THIS__",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "something :: points at",
    },
  },
]
`;

exports[`[getDotTriples] "__THIS__ :: is a :: note" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "note",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "__THIS__ :: is a",
    },
  },
]
`;

exports[`[getDotTriples] "A :: B :: C :: D" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "D",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "A :: B :: C",
    },
  },
]
`;

exports[`[getDotTriples] "Maria :: climbed :: the tree" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "the tree",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "Maria :: climbed",
    },
  },
]
`;

exports[`[getDotTriples] "arrays :: [banana, apple]" 1`] = `
Array [
  Object {
    "object": Object {
      "value": Array [
        Object {
          "entities": Object {},
          "value": "banana",
        },
        Object {
          "entities": Object {},
          "value": "apple",
        },
      ],
    },
    "predicate": Object {
      "entities": Object {},
      "value": "arrays",
    },
  },
]
`;

exports[`[getDotTriples] "date :: 2022-01-07 15:07" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {
        "dates": Array [
          Object {
            "duration": Object {
              "hours": 8,
              "minutes": 533,
            },
            "end": "2022-01-07T23:59:59.999+01:00",
            "start": "2022-01-07T15:07:00.000+01:00",
            "text": "2022-01-07 15:07",
            "tz": "Europe/Brussels",
            "unit": "time",
          },
        ],
      },
      "value": "2022-01-07 15:07",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "date",
    },
  },
]
`;

exports[`[getDotTriples] "external link :: <http://example.org>" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "<http://example.org>",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "external link",
    },
  },
]
`;

exports[`[getDotTriples] "external link :: http://example.org" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "http://example.org",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "external link",
    },
  },
]
`;

exports[`[getDotTriples] "internal link :: [[link]]" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {
        "internalLinks": Array [
          "link",
        ],
      },
      "value": "[[link]]",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "internal link",
    },
  },
]
`;

exports[`[getDotTriples] "is a :: [[Editor]], [[Jour]]" 1`] = `
Array [
  Object {
    "object": Object {
      "value": Array [
        Object {
          "entities": Object {},
          "value": "[Editor]]",
        },
        Object {
          "entities": Object {},
          "value": "[[Jour]",
        },
      ],
    },
    "predicate": Object {
      "entities": Object {},
      "value": "is a",
    },
  },
]
`;

exports[`[getDotTriples] "properties :: and values are separated by double colons" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "and values are separated by double colons",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "properties",
    },
  },
]
`;

exports[`[getDotTriples] "subject :: 'Quotes escape :: characters' :: object" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {},
      "value": "object",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "subject :: 'Quotes escape :: characters'",
    },
  },
]
`;

exports[`[getDotTriples] "vaccination dates :: 2022-01-07, next week" 1`] = `
Array [
  Object {
    "object": Object {
      "entities": Object {
        "dates": Array [
          Object {
            "duration": Object {},
            "end": null,
            "start": null,
            "text": "2022-01-07, next week",
            "tz": "Europe/Brussels",
            "unit": "time",
          },
        ],
      },
      "value": "2022-01-07, next week",
    },
    "predicate": Object {
      "entities": Object {},
      "value": "vaccination dates",
    },
  },
]
`;
